Packet++/src/GtpLayer.cpp

```diff
@@ -155,522 +155,521 @@ GtpV1Layer::GtpExtension GtpV1Layer::GtpExtension::createGtpExtension(uint8_t* d
-    }
-    else
-    {
-        // mark extension flags in the layer
-        header->extensionHeaderFlag = 1;
-        headerExtra->nextExtensionHeader = extensionType;
-    }
-
-    // create the extension data and return the extension object to the user
-    return GtpV1Layer::GtpExtension::createGtpExtension(
-        m_Data + offsetForNewExtension, 
-        m_DataLen - offsetForNewExtension, 
-        extensionType,
-        extensionContent);
-}
-
-GtpV1MessageType GtpV1Layer::getMessageType()
-{
-    gtpv1_header* header = getHeader();
-
-    if (header == NULL)
-    {
-        return GtpV1_MessageTypeUnknown;
-    }
-
-    return (GtpV1MessageType)header->messageType;
+	// get GTP header
+	gtpv1_header *header = getHeader();
+	if (header == NULL)
+	{
+		LOG_ERROR("Add extension failed: GTP header is NULL");
+		return GtpExtension();
+	}
+
+	size_t offsetForNewExtension = sizeof(gtpv1_header);
+
+	// if all flags are unset then create the GTP extra header
+	if (header->npduNumberFlag == 0 && header->sequenceNumberFlag == 0 && header->extensionHeaderFlag == 0)
+	{
+		if (!extendLayer(offsetForNewExtension, sizeof(gtpv1_header_extra)))
+		{
+			LOG_ERROR("Add extension failed: cannot extend layer");
+			return GtpExtension();
+		}
+		header = getHeader();
+	}
+
+	// get the extra header
+	gtpv1_header_extra *headerExtra = getHeaderExtra();
+	if (headerExtra == NULL)
+	{
+		LOG_ERROR("Add extension failed: extra header is NULL");
+		return GtpExtension();
+	}
+
+	offsetForNewExtension += sizeof(gtpv1_header_extra);
+
+	// find the last GTP header extension
+	GtpV1Layer::GtpExtension lastExt = getNextExtension();
+
+	// go over the GTP header extensions
+	while (!lastExt.getNextExtension().isNull())
+	{
+		// add ext total length to offset  
+		offsetForNewExtension += lastExt.getTotalLength();
+		lastExt = lastExt.getNextExtension();
+	}
+
+	// lastExt != null means layer contains 1 or more extensions
+	if (!lastExt.isNull())
+	{
+		// add ext total length to offset  
+		offsetForNewExtension += lastExt.getTotalLength();
+	}
+
+	// allocate extension space in layer (assuming extension length can only be 4 bytes)
+	if (!extendLayer(offsetForNewExtension, 4 * sizeof(uint8_t)))
+	{
+		LOG_ERROR("Add extension failed: cannot extend layer");
+		return GtpExtension();
+	}
+
+	// lastExt != null means layer contains 1 or more extensions
+	if (!lastExt.isNull())
+	{
+		// set the next header type in the last extension
+		lastExt.setNextHeaderType(extensionType);
+	}
+	else
+	{
+		// mark extension flags in the layer
+		header->extensionHeaderFlag = 1;
+		headerExtra->nextExtensionHeader = extensionType;
+	}
+
+	// create the extension data and return the extension object to the user
+	return GtpV1Layer::GtpExtension::createGtpExtension(
+		m_Data + offsetForNewExtension,
+		m_DataLen - offsetForNewExtension,
+		extensionType,
+		extensionContent);
+}
+
+GtpV1MessageType GtpV1Layer::getMessageType() const
+{
+	gtpv1_header *header = getHeader();
```

Comment:
please move the `*` back to its original place. This is the convention in the rest of the project

Reply:
Done